streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -224,47 +205,102 @@ private void addStandbyTasks() {
+            restoreConsumer.assign(updatedAssignment);
+        }
     }
 
     /**
      * Similar to shutdownTasksAndState, however does not close the task managers, in the hope that
-     * soon the tasks will be assigned again
+     * soon the tasks will be assigned again.
+     * @return list of suspended tasks
      * @throws TaskMigratedException if the task producer got fenced (EOS only)
      */
-    void suspendTasksAndState()  {
-        log.debug("Suspending all active tasks {} and standby tasks {}", active.runningTaskIds(), standby.runningTaskIds());
-
+    Set<TaskId> suspendActiveTasksAndState(final Collection<TopicPartition> revokedPartitions)  {
         final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        final Set<TaskId> revokedTasks = partitionsToTaskSet(revokedPartitions);
 
-        firstException.compareAndSet(null, active.suspend());
-        // close all restoring tasks as well and then reset changelog reader;
-        // for those restoring and still assigned tasks, they will be re-created
-        // in addStreamTasks.
-        firstException.compareAndSet(null, active.closeAllRestoringTasks());
-        changelogReader.reset();
+        log.debug("Suspending active tasks {}", revokedTasks);
 
-        firstException.compareAndSet(null, standby.suspend());
+        final List<TopicPartition> revokedChangelogs = new ArrayList<>();
 
-        // remove the changelog partitions from restore consumer
-        restoreConsumer.unsubscribe();
+        // suspend revoked active tasks and collect changelogs of any tasks that failed to suspend
+        firstException.compareAndSet(null, active.revokeTasks(revokedTasks, revokedChangelogs));
+
+        changelogReader.remove(revokedChangelogs);
+
+        // If the restore consumer is assigned still-restoring partitions, remove these from its assignment
+        if (!restoreConsumerAssignedStandbys) {
+            final Set<TopicPartition> updatedAssignment = new HashSet<>(restoreConsumer.assignment());
+            updatedAssignment.removeAll(revokedChangelogs);
+            restoreConsumer.assign(updatedAssignment);
+        }
 
         final Exception exception = firstException.get();
         if (exception != null) {
             throw new StreamsException(logPrefix + "failed to suspend stream tasks", exception);
         }
+        return active.suspendedTaskIds();
+    }
+
+    /**
+     * Closes active tasks as zombies, as these partitions have been lost and are no longer owned.
+     * @return list of lost tasks
+     */
+    Set<TaskId> closeLostTasks(final Collection<TopicPartition> lostPartitions) {
+        final Set<TaskId> zombieTasks = partitionsToTaskSet(lostPartitions);
+        log.debug("Closing lost tasks as zombies: {}", zombieTasks);
+
+        active.closeZombieTasks();
+        assignedActiveTasks.clear();
+        changelogReader.clear();
+        if (!restoreConsumerAssignedStandbys) {
```

Comment:
Not sure I understand this: we are assuming the parameter `lostPartitions` would always include all the active assigned partitions (which is true with the current protocol, but not necessarily always the case with future protocols right)? Also see my other comment: we can blindly remove the lost partitions since if they not assigned due to restoring, removal is a no-op.

Reply:
Do we envision there ever being a reason to "lose" only some subset of partitions? I felt it was reasonable to just clear everything as lost, but WDYT?

Reply:
I think it is safe to still rely on the passed in parameter to clear tasks than assuming it's always the whole set of partitions (though I agree it is the case indeed), since we do have them anyways

Reply:
Ack ok, will update

Reply:
But I think (at least for now, may change with future protocols of course) we should throw an exception if the lost tasks are not equal to all the assignedActiveTasks

Reply:
SG.